### ISSUE TITLE
mod_wsgi fails to build with Python 3.10: fatal error: node.h: No such file or directory

### ISSUE THREAD

**Issue by uhliarik** [GrahamDumpleton/mod_wsgi#687]
mod_wsgi files to build with Python 3.10.0a2:

```
+ /usr/bin/make -O -j2 V=1 VERBOSE=1
/usr/lib64/httpd/build/vendor-apxs -Wl,-export-symbols-regex -Wl,wsgi_module -c -I/usr/include/python3.10 -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -D_GNU_SOURCE -D_GNU_SOURCE -D_GNU_SOURCE  -Wc,-O2 -Wc,-flto=auto -Wc,-ffat-lto-objects -Wc,-fexceptions -Wc,-g -Wc,-grecord-gcc-switches -Wc,-pipe -Wc,-Wall -Wc,-Werror=format-security -Wc,-Wp,-D_FORTIFY_SOURCE=2 -Wc,-Wp,-D_GLIBCXX_ASSERTIONS -Wc,-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -Wc,-fstack-protector-strong -Wc,-specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -Wc,-m64 -Wc,-mtune=generic -Wc,-fasynchronous-unwind-tables -Wc,-fstack-clash-protection -Wc,-fcf-protection -Wc,-fno-strict-aliasing  src/server/mod_wsgi.c src/server/wsgi_*.c -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld  -L/usr/lib64 -L/usr/lib64 -L/usr/lib64/python3.10/config  -lpython3.10 -lcrypt -lpthread -ldl  -lutil -lm -lm
/usr/lib64/apr-1/build/libtool --silent --mode=compile gcc -prefer-pic -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -DLINUX -D_REENTRANT -D_GNU_SOURCE -pthread -I/usr/include/httpd  -I/usr/include/apr-1   -I/usr/include/apr-1  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-strict-aliasing -I/usr/include/python3.10 -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -D_GNU_SOURCE -D_GNU_SOURCE -D_GNU_SOURCE  -c -o src/server/mod_wsgi.lo src/server/mod_wsgi.c && touch src/server/mod_wsgi.slo
In file included from src/server/mod_wsgi.c:22:
src/server/wsgi_python.h:44:10: fatal error: node.h: No such file or directory
   44 | #include "node.h"
      |          ^~~~~~~~
compilation terminated.
```

According to https://docs.python.org/3.10/whatsnew/3.10.html :
```
Removed the parser module, which was deprecated in 3.9 due to the switch to the new PEG parser, 
as well as all the C source and header files that were only being used by the old parser, including node.h, 
parser.h, graminit.h and grammar.h.
```

**Comment by encukou** [comment]
A bit of background:
- Python's old LL(1) parser is gone; it was limiting Python's syntax too much. (For a small example, in 3.10 you can finally [use parentheses with `with`](https://docs.python.org/3.10/whatsnew/3.10.html#parenthesized-context-managers), though I imagine it might not look like a win to everyone.)
- The new parser is a backtracking one, and needs the entire source in memory as a `char` array.
- New Python APIs generally use Python "file-like objects" rather than `FILE*` as the abstraction of choice.

So, `mod_wsgi` will unfortunately need to read an entire file to a `char` array. This is somewhat fiddly in C; does Apache/mod_wsgi happen to have an existing helper for it?

When you have a `char*`, call [Py_CompileString](https://docs.python.org/3.10/c-api/veryhigh.html#c.Py_CompileString) to get a code object, rather than go through the intermediate `node`.

**Comment by GrahamDumpleton** [comment]
No helper, but I wouldn't say it is particularly fiddly.

**Comment by GrahamDumpleton** [comment]
Note that Python 3.10 support is added for mod_wsgi 4.9.0 but could do with some people testing the develop version from GitHub of it before being released to make sure works and that all the changes also don't cause memory issues with older Python versions since API changes also affected old versions.

**Comment by encukou** [comment]
Fedora 35 (which will go into [beta in 3 weeks](https://fedorapeople.org/groups/schedule/f-35/f-35-key-tasks.html)) will have Python 3.10. The `mod_wsgi` package includes the patch from #688 so it can build. I know that at least the [FreeIPA](https://www.freeipa.org/page/Main_Page) team runs their tests with it.
Could you publish a beta or release candidate? That would make it easier to get the other changes in as well, and hopefully it would make it easier for others to test too.

The gist of #688 is now also in [Python's release notes](https://docs.python.org/3.10/whatsnew/3.10.html#changes-in-the-c-api). If  anyone else reports an issue with it, I'll let you know.

**Comment by GrahamDumpleton** [comment]
If there were a new release of mod_wsgi by end of this week, would you switch to it or go with the old version you have? Work gave us a week off so been catching up on lots of mod_wsgi stuff which had been accumulated. Almost ready for release, but was worried about how much Python 3.10 changes were tested.

**Comment by GrahamDumpleton** [comment]
FWIW, can see release notes in:

* https://github.com/GrahamDumpleton/mod_wsgi/blob/develop/docs/release-notes/version-4.9.0.rst

Right now just finishing up some stuff with ``mod_wsgi-standalone`` package to update it to newer Apache version and drop some old crud related to Heroku and OpenShift from ``setup.py`` file.

**Comment by GrahamDumpleton** [comment]
BTW, 4.9.0 also has some changes related to setting ``LD_RUN_PATH`` which I think Fedora wanted. Just not mentioned in released notes.

**Comment by encukou** [comment]
> If there were a new release of mod_wsgi by end of this week, would you switch to it or go with the old version you have?

It's not up to me as I'm not the Fedora maintainer, but I [asked](https://bugzilla.redhat.com/show_bug.cgi?id=1960695#c3).

**Comment by GrahamDumpleton** [comment]
Will not be a beta, but would be the actual release. And is 4.9.0 not 3.9.0 as you mentioned in other release. I could release it right now but just been seeing if there is anything left in issues list I can address after I did a big cleanup and closed out lots of issues.

**Comment by GrahamDumpleton** [comment]
Hmmm, I may have just found something that the Python 3.10 changes break :-(

**Comment by GrahamDumpleton** [comment]
Have released mod_wsgi 4.9.0 now.

**Comment by hroncok** [comment]
A Fedora PR is here: https://src.fedoraproject.org/rpms/mod_wsgi/pull-request/12
